internal/validation/apiextensions/v1/composition/handler.go

```diff
@@ -74,103 +67,91 @@ func SetupWebhookWithManager(mgr ctrl.Manager, options controller.Options) error
-		}
-		return admission.Denied(err.Error()).WithWarnings(warns...)
-	case admissionv1.Delete:
-		return admission.Allowed("")
-	case admissionv1.Connect:
-		return admission.Errored(http.StatusBadRequest, errors.New("unexpected operation"))
-	default:
-		return admission.Errored(http.StatusBadRequest, errors.New("unexpected operation"))
-	}
+type validator struct {
+	reader  client.Reader
+	options controller.Options
 }
 
-func validationResponseFromStatus(allowed bool, status metav1.Status) admission.Response {
-	resp := admission.Response{
-		AdmissionResponse: admissionv1.AdmissionResponse{
-			Allowed: allowed,
-			Result:  &status,
-		},
+// ValidateCreate validates a Composition.
+func (v *validator) ValidateCreate(ctx context.Context, obj runtime.Object) (admission.Warnings, error) { //nolint:gocyclo // Currently only at 11
+	comp, ok := obj.(*v1.Composition)
+	if !ok {
+		return nil, errors.New(errNotComposition)
 	}
-	return resp
-}
 
-// Validate validates the Composition by rendering it and then validating the rendered resources.
-func (h *handler) Validate(ctx context.Context, comp *v1.Composition) (warns []string, err error) {
 	// Validate the composition itself, we'll disable it on the Validator below
-	var validationErrs field.ErrorList
-	warns, validationErrs = comp.Validate()
+	warns, validationErrs := comp.Validate()
 	if len(validationErrs) != 0 {
 		return warns, apierrors.NewInvalid(comp.GroupVersionKind().GroupKind(), comp.GetName(), validationErrs)
 	}
 
-	if !h.options.Features.Enabled(features.EnableAlphaCompositionWebhookSchemaValidation) {
+	if !v.options.Features.Enabled(features.EnableAlphaCompositionWebhookSchemaValidation) {
 		return warns, nil
 	}
 
 	// Get the composition validation mode from annotation
 	validationMode, err := comp.GetValidationMode()
 	if err != nil {
-		return warns, xperrors.Wrap(err, "cannot get validation mode")
+		return warns, errors.Wrap(err, errValidationMode)
 	}
 
-	// Get all the needed CRDs, Composite Resource, Managed resources ... ? Error out if missing in strict mode
-	gkToCRD, errs := h.getNeededCRDs(ctx, comp)
-	// if we have errors, and we are in strict mode or any of the errors is not a , return them
+	// Get all the needed CRDs, Composite Resource, Managed resources ... ?
+	// Error out if missing in strict mode
+	gkToCRD, errs := v.getNeededCRDs(ctx, comp)
+	// if we have errors, and we are in strict mode or any of the errors is not
+	// a , return them
 	if len(errs) != 0 {
 		if validationMode == v1.CompositionValidationModeStrict || containsOtherThanNotFound(errs) {
-			return warns, xperrors.Errorf("there were some errors while getting the needed CRDs: %v", errs)
+			// TODO(negz): Do we really need to return all the errors? Typically
+			// we just return the first one we hit. This works well with
+			// errors.Wrap. Is there extra value in returning them all here as a
+			// special case?
```

Comment:
That seems like a good idea. I think first step would be to surface that function in `pkg/errors` in crossplane-runtime. I'll update the TODO.

I can see in this case where we're validating an object that it's more worth it than usual to try process the whole thing and return all errors so that you don't have to iterate through each error case to discover them.